File: src/types/pattern.ts
export type PatternPieceId = string;

export type Point2D = {
  x: number;
  y: number;
};

export type CurveSegment
  = { type: "Line"; end: Point2D }
    | { type: "QuadraticBezier"; control: Point2D; end: Point2D }
    | { type: "CubicBezier"; control1: Point2D; control2: Point2D; end: Point2D }
    | {
      type: "Arc";
      center: Point2D;
      radius: number;
      start_angle: number;
      end_angle: number;
    };

export type PieceMetadata = {
  fabric_type: string | null;
  cut_quantity: number;
  mirror: boolean;
  notes: string;
};

export type PatternPieceData = {
  id: PatternPieceId;
  name: string;
  origin: Point2D;
  outline: CurveSegment[];
  grain_line: [Point2D, Point2D] | null;
  seam_allowance_mm: number;
  notches: Point2D[];
  internal_lines: CurveSegment[][];
  metadata: PieceMetadata;
};

File: src-tauri/src/types/pattern.rs
use serde::{Deserialize, Serialize};

pub type PatternPieceId = String;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Point2D {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum CurveSegment {
    Line {
        end: Point2D,
    },
    QuadraticBezier {
        control: Point2D,
        end: Point2D,
    },
    CubicBezier {
        control1: Point2D,
        control2: Point2D,
        end: Point2D,
    },
    Arc {
        center: Point2D,
        radius: f64,
        start_angle: f64,
        end_angle: f64,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PieceMetadata {
    pub fabric_type: Option<String>,
    pub cut_quantity: u32,
    pub mirror: bool,
    pub notes: String,
}

impl Default for PieceMetadata {
    fn default() -> Self {
        Self {
            fabric_type: None,
            cut_quantity: 1,
            mirror: false,
            notes: String::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PatternPieceData {
    pub id: PatternPieceId,
    pub name: String,
    pub origin: Point2D,
    pub outline: Vec<CurveSegment>,
    pub grain_line: Option<(Point2D, Point2D)>,
    pub seam_allowance_mm: f64,
    pub notches: Vec<Point2D>,
    pub internal_lines: Vec<Vec<CurveSegment>>,
    pub metadata: PieceMetadata,
}
